### ISSUE TITLE
some performance results of envoy's different versions

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#19103]
I did some simple tests today and found that envoy's performance seems to be getting worse and worse. Although I know features should always come at some cost. But those costs seem to be too much.

And although I know that performance is not the first goal of Envoy. But with continuous enhancement of Envoy, the performance seems to degrade too quickly. Here are some simple results of **single envoy worker**:

| Version | QPS      |
| ------- | -------- |
| v1.12.4 | 26921.48 |
| v1.13.3 | 25182.18 |
| v1.14.7 | 23732.31 |
| v1.15.5 | 21010.66 |
| v1.16.5 | 19116.81 |
| v1.17.4 | 17804.78 |
| v1.18.4 | 16953.67 |
| v1.19.1 | 16046.59 |
| v1.20.0 | 15949.65 |

I think we should at least prevent it from further deterioration, and at the same time find ways to optimize it.

Here is my config yaml:

```
static_resources:
  listeners:
    - address:
        socket_address:
          address: 0.0.0.0
          port_value: 9090
      filter_chains:
        - filters:
            - name: envoy.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                codec_type: auto
                stat_prefix: ingress_http
                route_config:
                  name: local_route
                  virtual_hosts:
                    - name: host-one # prefix route
                      domains:
                        - "*"
                      routes:
                        - match:
                            prefix: "/"
                          route:
                            cluster: httpbin
                http_filters:
                  - name: envoy.router
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
                      dynamic_stats: false
                      suppress_envoy_headers: true
  clusters:
    - name: httpbin
      connect_timeout: 5s
      type: strict_dns
      lb_policy: round_robin
      load_assignment:
        cluster_name: httpbin
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: localhost
                      port_value: 8080

admin:
  access_log_path: "/dev/null"
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 20000

```

**Comment by wbpcode** [comment]
cc @mattklein123  @rojkov

**Comment by zhxie** [comment]
Have you ever profiled Envoy with tools like perf, so that we can find some hotspots from the result?

**Comment by wbpcode** [comment]
> Have you ever profiled Envoy with tools like perf, so that we can find some hotspots from the result?

I will do it this weekend.

**Comment by rojkov** [comment]
This feels like a duplicate of #13412.

**Comment by wbpcode** [comment]
> This feels like a duplicate of #13412.

Yep. Looks like the degrading never stops.

**Comment by wbpcode** [comment]
I only have two different binary files with code symbols. Considering that there has been a huge gap in performance between them, I think it is enough as a reference. 


v1.17.4:
https://drive.google.com/file/d/1aZuC54PmIXsQu7k88jBeU-HENOEiQXgT/view?usp=sharing

v1.12.2:
https://drive.google.com/file/d/1o6kV8T2J5nCs3m7lHaXSdXaudjzNJ9Sv/view?usp=sharing

**Comment by wbpcode** [comment]
There are no obvious hotspots from the flame graph, just a very uniform slowdown. More encapsulation and abstraction gradually reduces the performance of Envoy.

**Comment by jmarantz** [comment]
I'm not sure if I buy that encapsulation and abstraction are likely to cause that much slowdown. I glanced at one of the flame-graphs and it's hard to know what to improve there.

It might help to capture some perf-graphs from an instrumented binary that can provide more detail on what's going on. The purely sampled view we get from these flame-graphs might (for example, as a guess) hide some effects of changes in the way we make the networking system calls. E.g. we spend a lot of time in the 1.20 flamegraph in the kernel. Are we making larger numbers of calls for smaller chunks of data, for any reason?

It would be nice to get a controlled repro of this, using NIghthawk preferably, and then run envoy under `callgrind` or compiled with various perf  tools, so we can see call-counts for various functions. Using this we could compare how we are structuring our system calls to the earlier version of Envoy.

**Comment by wbpcode** [comment]
> I'm not sure if I buy that incremental changes in encapsulation and abstraction are likely to cause that much slowdown. I glanced at one of the flame-graphs and it's hard to know what to improve there.
> 
> It might help to capture some perf-graphs from an instrumented binary that can provide more detail on what's going on. The purely sampled view we get from these flame-graphs might (for example, as a guess) hide some effects of changes in the way we make the networking system calls. E.g. we spend a lot of time in the 1.20 flamegraph in the kernel. Are we making larger numbers of calls for smaller chunks of data, for any reason?
> 
> It would be nice to get a controlled repro of this, using NIghthawk preferably, and then run envoy under `callgrind` or compiled with various perf  tools, so we can see call-counts for various functions. Using this we could compare how we are structuring our system calls to the earlier version of Envoy.

I will try to do more investigate.

**Comment by mattklein123** [comment]
Related to @jmarantz comment, over time we have generally moved to secure by default configuration. This almost universally is at odds with "horse race" benchmarks. So we will need to very carefully tease apart any differences. The work is useful but it's a lot more complicated than just looking at flame graphs. I also agree with @jmarantz that trying to run a reproducer under something like cachegrind will be a lot more informative.

**Comment by jmarantz** [comment]
Just to point this out: Matt suggests cachegrind and I suggested callgrind: they are related and both very useful:
 * they both are generated using the valgrind infrastructure (`--tool=callgrind` vs `--tool=cachegrind`)
 * cachegrind has a lot more detail of the simulated processor cache effects of each line of code
 * callgrind has more detail about time spent and call-counts at a function-level 
 * they both benefit from an OptDebug compilation. I use ` --compilation_mode=opt --cxxopt=-g --cxxopt=-ggdb3` but those options might be dated as I haven't been able to do much direct technical work lately
 * they both generate reports that can be visualized with `kcachegrind`
 * they both run much slower than real time and I wouldn't run production traffic through them. But it should be no problem for nighthawk or any other synthetic load tool

**Comment by wbpcode** [comment]
@mattklein123 @jmarantz Thanks very much for all your suggestions. I will put more effort into tracking and investigating this issue recently.🌷 

At present, I still subjectively think that it may caused by a large number of small adjustments. For example, #13423 introduces a small update to replace all white space chars in the response code details which will bring some minor new overhead. Assuming there are many similar PRs, the accumulation of these small overheads will still have a large enough impact. And because they are very scattered, it may also be difficult to locate. 

But this is just my personal guess. We still need more research to identify the problem and try to solve it.

Of course, if it is just because of some more secure default configuration, then at most only the documentation needs to be updated.

**Comment by wbpcode** [comment]
2021/11/27 callgrind.out file of v1.12.2 vs v1.20.x (100, 000 HTTP1.1 request with wrk).

https://drive.google.com/drive/folders/1EWTjixvN43O8u24a_rJF8S6A4ePvFwC1?usp=sharing

![v1.20.x](https://user-images.githubusercontent.com/12389633/143672687-95f0d73b-2257-44a2-9467-76eb653be109.png)

![v1.12.2](https://user-images.githubusercontent.com/12389633/143672706-10cc6065-9123-4522-a152-3beb9b63d3ff.png)


First point: `encodeFormattedHeader` introduced ~3% external CPU overhead.
Related PR: #9825

**Comment by jmarantz** [comment]
Great -- would you be able to supply the parameters passed to Nighthawk (preferably) or whatever tool you were using to generate the load for these traces?

Thanks!

**Comment by jmarantz** [comment]
This data is great! It really shows the profiles look very different. Just sorting by 'self' and comparing the two views provides a ton of possible issues to go explore.

Did you use an "OptDebug" build or was this a stock build or something?

**Comment by wbpcode** [comment]
> 

I used simple `hey` to generate 100, 000 http1 request. Here is the command.

`hey -n 100000 -c 20 http://localhost:9090`.

**Comment by jmarantz** [comment]
I had not heard of `hey`. Is that this? https://github.com/rakyll/hey

**Comment by wbpcode** [comment]
> Did you use an "OptDebug" build or was this a stock build or something?

I used the no stripped binary that build with ` ci/do_ci.sh bazel.release`. It finally worked, but I'm not sure it is OptDebug mode.

**Comment by wbpcode** [comment]
> I had not heard of `hey`. Is that this? https://github.com/rakyll/hey

Yes, I am generally used to using wrk, but it seems that wrk cannot generate a fixed amount of load.

**Comment by jmarantz** [comment]
Nighthawk (https://github.com/envoyproxy/nighthawk) is really what we want to converge to, as you get explicit control of open/closed loop, ability to generate http2, async vs concurrent, an indication of how many requests succeeded/failed, etc.

By "fixed amount of load" does this mean a finite number of requests? Or a fixed rate of requests?

The docker images are definitely *not* in OptDebug mode :) They are probably simply optimized, which is OK, but we'll have a lot less details on call-stack, where in functions time is being spent, etc.

**Comment by wbpcode** [comment]
> Nighthawk (https://github.com/envoyproxy/nighthawk) is really what we want to converge to, as you get explicit control of open/closed loop, ability to generate http2, async vs concurrent, an indication of how many requests succeeded/failed, etc.

Thanks, I will try it in the coming test.

> The docker images are definitely not in OptDebug mode :) They are probably simply optimized, which is OK, but we'll have a lot less details on call-stack, where in functions time is being spent, etc.

I see. Thanks. 🌷  I will try to build new binary with you suggested compile args and do some more investigates. But recently I only have enough time on weekends.

**Comment by jmarantz** [comment]
No worries -- the data you supplied is great. We'll definitely want to repro with NH though so we understand what we are comparing :)

My suspicion is that you've found something real, and based on the traces I looked at, there were changes in either tcmalloc's implementation, the way we allocate buffers, or both. The `encodeFormattedHeader` hot-spot probably deserves a quick look also as that is a relatively recent feature. Did you wind up turning that on?

What did you have running on localhost:8080? An Apache server or something?

**Comment by wbpcode** [comment]
> What did you have running on localhost:8080? An Apache server or something?

A multi-processes Nginx that will return simple short string directly.

> Did you wind up turning that on?

In fact, nope. Nevertheless, the new version of `encodeFormatedHeader/encodeHeader` still introduces more overhead. The reason is that `encodeHeaders` call `encodeFormatedHeader` several times and write some fine-grained data (a character, a space, a header name, etc.) to the buffer.
And the new version does not use a local cache to speed up the writing of these fine-grained data.

I've created a PR #19115 to try to solve this problem.

**Comment by hobbytp** [comment]
@wbpcode do you mean the issue only exists in HTTP1 or you only test it by using HTTP1, I ask this because I found in your PR  #19115, you only fix HTTP1 codes, thanks for clarification.

**Comment by wbpcode** [comment]
> 

I only test it by using HTTP1. And codec is not the only problem. We need more investigations and PRs to continue to alleviate this problem.

**Comment by wbpcode** [comment]
> @wbpcode do you mean the issue only exists in HTTP1 or you only test it by using HTTP1, I ask this because I found in your PR #19115, you only fix HTTP1 codes, thanks for clarification.

I only test it by using HTTP1. And codec is not the only problem. We need more investigations and PRs to continue to alleviate this problem.

**Comment by lambdai** [comment]
What's the envoy cpu usage between 1.12 and 1.20? 
I asked because I remember some event handle is pushed to the next poller cycle.
It means envoy trades latency for fairness. 
Now that the benchmark client maintains the 20 active req/resp, it's likely the cpu usage at envoy is lower in 1.20

**Comment by wbpcode** [comment]
@lambdai 

When I tested the throughput, both 1.12 and 1.20 used a full core. When I collected callgrind data later, because it was executed in valgrind, I didn't pay attention to the cpu situation. Moreover, when we analyze callgrind data, my main concern is the change in the proportion of overhead of different functions.

**Comment by moderation** [comment]
Via https://news.ycombinator.com/item?id=29457579 there are some good examples of projects that track performance over time:

- https://ziglang.org/perf/
- https://perf.rust-lang.org/
- https://llvm-compile-time-tracker.com/

**Comment by rgs1** [comment]
Not to hijack this thread, but this might be possibly related: on our end we recently started looking at `slow_alloc()` dominating CPU time as seen in our view of Envoy via `perf top`:

![Screen Shot 2021-12-10 at 2 33 02 PM](https://user-images.githubusercontent.com/1349891/145631251-d20f629e-bc9b-439b-bbdf-6a916abf575f.png)

cc: @fishcakez

**Comment by VA-PLE** [comment]
Thank you for this information. We use envoy in high-load projects. It's amazing to see such a drop in performance in the new version.

I hope the test will be replenished with new versions of the envoy and performance will improve over time

**Comment by wbpcode** [comment]
@VA-PLE We are working towards this.

**Comment by jiangdongzi** [comment]
I test and test and get another result:
1.  Using wrk test: v1.7 uses much more cpu and memory and gives better performance. Both latency and tps
2. Using wrk test: If envoy is the bottleneck, ie. envoy cpu using hit 100%, then v1.3 is better than v1.7. Both latency and tps.
3. Using nighthawk: Whatever I do, v1.3 is better than v1.7

**Comment by jmarantz** [comment]
Are 1.3 and 1.7 versions of something other than Envoy? Those seem like very low version numbers.

I think it probably makes sense to do more benchmarking on HEAD once #19115 lands. Then we should capture more traces to find the next bottleneck.

**Comment by jiangdongzi** [comment]
> Are 1.3 and 1.7 versions of something other than Envoy? Those seem like very low version numbers.
> 
> I think it probably makes sense to do more benchmarking on HEAD once #19115 lands. Then we should capture more traces to find the next bottleneck.

Sorry, please just ignore my previous commont. When I use wrk test, I always test v1.3 first(The cache may be cold)  This may gave me wrong answer. Just now, I retest, the result is the tps between v1.3 and v1.7 is very close;  and from latency aspect, the v1.7 is slightly better.  But v1.7 uses 20% much more cpu than v1.3.  But the nighthawk result is always that v1.3 is 20% better than v1.7. 
I'll continue testing.

**Comment by wbpcode** [comment]
It would be better to compare some new versions. `v1.3` and `v1.7` look like too old.

**Comment by jmarantz** [comment]
Yeah they way too old to be useful Envoy versions. I suspect @jiangdongzi is referring to an Istio version number or some other packaging of Envoy.

In any case, there are a lot of complexities to benchmarking Envoy well, and it's worth reading up on:
 * https://github.com/envoyproxy/envoy/blob/main/bazel/PPROF.md
 *  I thought @htuch wrote a white-paper on best practices for benchmarking Envoy but I can't find it -- maybe he can follow up with a pointer.

**Comment by jmarantz** [comment]
I found the material @htuch wrote but it's in a formal doc, not a white paper: https://www.envoyproxy.io/docs/envoy/latest/faq/performance/how_to_benchmark_envoy

**Comment by wbpcode** [comment]
Some insteresting result from perf. The `readDisable` calls from HTTP1 codec result in ~7% cpu overhead.

![image](https://user-images.githubusercontent.com/12389633/153335488-d34a7140-4c7d-4d34-8457-800b925b0841.png)

![image](https://user-images.githubusercontent.com/12389633/153335901-694e8163-21e4-464e-b4fa-cf35209b4908.png)

**Comment by ztgoto** [comment]
I am on a machine with Intel(R) Xeon(R) CPU E5-2620 v4 @ 2.10GHz, --concurrency=1, all on one host, the proxy qps can only reach about 10k, and nginx (worker_processes 1, qps 27K) is a huge difference. And as the number of worker threads increases, the gap gets bigger and bigger. We are doing api gateway technology selection， Have you tested the performance comparison between envoy and nginx under the same environmental conditions?

**Comment by zhxie** [comment]
@ztgoto I have tested Envoy (envoy.filters.http.router) and Nginx (proxy_pass) as HTTP proxy, with both in single thread and have 64 conns loaded by fortio. Envoy has more than 10% performance gain compared to Nginx in RPS. Can you show example configurations in your case?

**Comment by ztgoto** [comment]
> @ztgoto I have tested Envoy (envoy.filters.http.router) and Nginx (proxy_pass) as HTTP proxy, with both in single thread and have 64 conns loaded by fortio. Envoy has more than 10% performance gain compared to Nginx in RPS. Can you show example configurations in your case?

envoy
```
admin:
  access_log_path: /dev/null
  address:
    socket_address: {address: 0.0.0.0, port_value: 7777}

static_resources:
  listeners:
  - name: listener
    address:
      socket_address: {address: 0.0.0.0, port_value: 9104}
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: http_test
          codec_type: AUTO
          generate_request_id: false
          route_config:
            name: route
            virtual_hosts:
            - name: test
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                route:
                  cluster: nginx
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
              dynamic_stats: false
  clusters:
  - name: "nginx"
    connect_timeout: "5s"
    type: STRICT_DNS
    lb_policy: ROUND_ROBIN
    circuit_breakers:
      thresholds:
      - priority: DEFAULT
        max_connections: 1000000000
        max_pending_requests: 1000000000
        max_requests: 1000000000
        max_retries: 0
      - priority: HIGH
        max_connections: 1000000000
        max_pending_requests: 1000000000
        max_requests: 1000000000
        max_retries: 0
    load_assignment:
      cluster_name: nginx
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: 192.168.1.100
                port_value: 9101

stats_config:
  stats_matcher:
    reject_all: true
```


nginx
```
worker_processes  1;


events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;

    access_log off;

    sendfile        on;

    keepalive_timeout  65;


    upstream test {
      keepalive 50;
      server 192.168.1.100:9101;

    }

    server {
      listen 9103;
      server_name 192.168.4.119;
      location / {
        proxy_http_version 1.1;
        proxy_set_header Connection "";
        proxy_pass http://test;
      }
    }


}
```

**Comment by jmarantz** [comment]
@krajshiva @yanavlasov here's some anecdotes about performance gaps between nginx and envoy.

**Comment by jmarantz** [comment]
@ztgoto PTAL @ https://www.envoyproxy.io/docs/envoy/latest/faq/performance/how_to_benchmark_envoy

In particular, can you disable circuit-breaking and re-run your benchmark (per recommendations in the benchmarking doc above). It's not clear to me if that's a bottleneck in your case, but you have not configured circuit-breaking on the nginx side so it seems appropriate to keep the behavior consistent.

**Comment by KBaichoo** [comment]
@ztgoto What HTTP protocol is being used? HTTP/1, or HTTP/2? In particular for HTTP2  in https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#config-core-v3-http2protocoloptions
`max_concurrent_streams` defaults to 2147483647 which I've seen this skew some benchmarks prior.

**Comment by ztgoto** [comment]
> @ztgoto PTAL @ https://www.envoyproxy.io/docs/envoy/latest/faq/performance/how_to_benchmark_envoy
> 
> In particular, can you disable circuit-breaking and re-run your benchmark (per recommendations in the benchmarking doc above). It's not clear to me if that's a bottleneck in your case, but you have not configured circuit-breaking on the nginx side so it seems appropriate to keep the behavior consistent.
> 
> Also, did you build your own Envoy from source? Or use a pre-built package? If you built it yourself did you use `-c opt`?

@jmarantz  I use envoyproxy/envoy:v1.22.2 (--network host) for testing. The config file is as stated above, I don't know if there is anything wrong.

envoy
```
wrk -t 8 -c 32  -d 60s --latency 'http://127.0.0.1:9104/hello'
Running 1m test @ http://127.0.0.1:9104/hello
  8 threads and 32 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.76ms  378.73us   6.53ms   66.32%
    Req/Sec     1.45k    65.16     2.12k    71.77%
  Latency Distribution
     50%    2.69ms
     75%    3.02ms
     90%    3.33ms
     99%    3.67ms
  693601 requests in 1.00m, 156.77MB read
Requests/sec:  11553.84
Transfer/sec:      2.61MB


   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                          
160976 101       20   0 2247104  54512  19432 S 100.3  0.1   2:13.85 envoy                                                            
 32221 root      20   0  706612   8384   1160 S  32.8  0.0   0:04.63 wrk  
```

nginx
```
wrk -t 8 -c 32  -d 60s --latency 'http://127.0.0.1:9103/hello'
Running 1m test @ http://127.0.0.1:9103/hello
  8 threads and 32 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.12ms  107.53us  13.65ms   78.66%
    Req/Sec     3.56k   202.05     4.44k    76.00%
  Latency Distribution
     50%    1.12ms
     75%    1.18ms
     90%    1.24ms
     99%    1.37ms
  1701221 requests in 1.00m, 379.64MB read
Requests/sec:  28351.42
Transfer/sec:      6.33MB

   PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                          
 58685 nobody    20   0   46588   2752   1116 R 100.0  0.0   0:10.66 nginx                                                            
 60538 root      20   0  701960   3728   1136 S  89.4  0.0   0:09.88 wrk
```

**Comment by scheruku-in** [comment]
Hi, 
We have upgraded our envoy version which is ~5 years old to latest v1.25.3. We have executed few perf tests and noticed that there seems to be 60% increase in CPU and 5x increase in memory compared to earlier version of envoy with same test. CPU went up-to 400m and memory was ~120M. Would some one please review the flame graph and share the findings if this can be further optimized? Thanks in advance.

Flame graph:
https://github.com/scheruku-in/Envoy_Perf_Test/blob/main/envoy_highcpu.svg


Envoy config:

```
{
  "listeners": [
    {
      "address": "tcp://0.0.0.0:17600",
      "ssl_context": {
        "ca_cert_file": "envoycacert.pem",
        "cert_chain_file": "cacertchain.pem",
        "private_key_file": "key.pem",
        "alpn_protocols": "h2,http/1.1",
        "alt_alpn_protocols": "http/1.1"
      }, 
      "filters": [
        {
          "type": "read",
          "name": "http_connection_manager",
          "config": {
            "access_log": [
              {
                "format": "[%START_TIME%]  \"%REQ(X-FORWARDED-FOR)%\" - \"%REQ(USER-AGENT)%\"  \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %BYTES_RECEIVED% %BYTES_SENT%  x-request-id = \"%REQ(X-REQUEST-ID)%\" x-global-transaction-id = \"%REQ(X-GLOBAL-TRANSACTION-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" \"%UPSTREAM_CLUSTER%\"    rt=\"%DURATION%\" uct=\"%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%\" \n",
                "path": "/dev/stdout"
              }],
            "generate_request_id": true,
            "codec_type": "auto",
            "idle_timeout_s": 120,
            "stat_prefix": "ingress_http",
            "route_config": {
              "virtual_hosts": [
                {
                  "name": "service",
                  "domains": ["*"],
                  "require_ssl": "all",
                  "routes": [
                    {
                      "timeout_ms": 120000,
                      "retry_policy": {
                          "retry_on": "gateway-error,connect-failure",
                          "num_retries": 120
                          },
     		          "prefix": "/",
     		          "cluster_header" : "<cluster-header>"
                    }
                  ]
                }
              ]
            },
            "filters": [
              {
                "type": "decoder",
                "name": "router",
                "config": {}
              }
            ]
          }
        }
      ]
    }
  ],
  "admin": {
    "access_log_path": "/dev/stdout",
    "address": "tcp://127.0.0.1:8001"
  },
  "cluster_manager": {
    "clusters": [
    ],
    "cds": {
      "cluster": {
        "name": "cds",
        "connect_timeout_ms": 120000,
        "type": "strict_dns",
        "lb_type": "round_robin",
        "hosts": [
          {
            "url": "tcp://localhost:8081"
          }
        ]
      },
      "refresh_delay_ms": 100
    },
    "outlier_detection": {
      "event_log_path": "/dev/stdout"
    }
  }
}
```


Thanks in advance.

**Comment by howardjohn** [comment]
Update from 2024:

| Version | QPS   |
| ------- | ----- |
| 1.13    | 65741 |
| 1.14    | 60302 |
| 1.15    | 52966 |
| 1.16    | 49947 |
| 1.17    | 45257 |
| 1.18    | 42121 |
| 1.19    | 41800 |
| 1.20    | 39946 |
| 1.21    | 39649 |
| 1.22    | 47829 |
| 1.23    | 45302 |
| 1.24    | 42135 |
| 1.25    | 41102 |
| 1.26    | 39932 |
| 1.27    | 41539 |
| 1.28    | 36995 |
| 1.29    | 37273 |
| 1.30    | 37962 |

<details>

```
crane-get-file () {
  crane export $1 - | tar -Oxf - $2
}

for i in {13..30}; do
  crane-get-file envoyproxy/envoy:v1.$i-latest usr/local/bin/envoy > envoy-$i
  chmod +x envoy-$i
done

for i in {13..30}; do
  echo "STARTING $i"
  ./envoy-$i -c config.yaml --concurrency 1 --disable-hot-restart -l off &
  p=$!
  sleep .1
  benchtool -q 0 -d 10 localhost:9090#envoy-1.$i >> res
  kill -9 $p
  echo "ENDING $i"
done
```

</details>

**Comment by ztgoto** [comment]
2024:
Returns the results directly,
The stress testing tool uses wrk
version:
envoy：1.30.4
nginx: 1.26.1
config:

envoy
`./envoy --concurrency 1 -c ./envoy.yaml`
```yaml
static_resources:
  listeners:
  - name: listener
    address:
      socket_address: {address: 0.0.0.0, port_value: 9101}
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: http_test
          codec_type: AUTO
          generate_request_id: false
          route_config:
            name: route
            virtual_hosts:
            - name: test
              domains: ["*"]
              routes:
              - match: { prefix: "/" }
                direct_response:
                  status: 200
                  body:
                    inline_string: "{\"message\":\"hello\"}"
                response_headers_to_add:
                - header:
                    key: "Content-Type"
                    value: "application/json"
              #  route:
              #    cluster: auth
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
              dynamic_stats: false

stats_config:
  stats_matcher:
    reject_all: true
```

nginx
```
worker_processes  1;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;

    access_log off;

    sendfile        on;

    keepalive_timeout  65;


    server {
        listen       9101;
        server_name  127.0.0.1;


        location / {
            default_type application/json;
            return 200 '{"message":"hello"}';
        }

        location = /50x.html {
            root   html;
        }

    }


}
```

result:

envoy:
```
./wrk -t 8 -c 1000 -d 60s 'http://127.0.0.1:9101'
Running 1m test @ http://127.0.0.1:9101
  8 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    69.34ms    7.74ms 123.77ms   64.42%
    Req/Sec     1.81k   521.16     4.62k    38.42%
  864304 requests in 1.00m, 117.05MB read
Requests/sec:  14382.08
Transfer/sec:      1.95MB
```

nginx:
```
./wrk -t 8 -c 1000 -d 60s 'http://127.0.0.1:9101'
Running 1m test @ http://127.0.0.1:9101
  8 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    39.28ms   99.20ms   2.00s    95.30%
    Req/Sec     6.09k   505.22     7.36k    72.21%
  2908817 requests in 1.00m, 479.91MB read
  Socket errors: connect 0, read 4672, write 0, timeout 13
Requests/sec:  48404.51
Transfer/sec:      7.99MB
```

Flame diagram
![envoy-1.30.4](https://github.com/envoyproxy/envoy/assets/16531243/a31ed07d-0b51-4703-9e05-b16c3d01841e)

![nginx-1.26.1](https://github.com/envoyproxy/envoy/assets/16531243/c6039a28-e373-415c-a8af-0f64525767fa)

**Comment by zhxie** [comment]
> 2024: Returns the results directly, The stress testing tool uses wrk version: envoy：1.30.4 nginx: 1.26.1 config:

I have benchmarked Envoy with `direct_response` once with Fortio **with keep-alive disabled**. I noticed that Envoy does not close the connection immediately after sending the response, which leads to performance degradation. Your scenario differs from mine, but `direct_response` isn't commonly used and may not be fully optimized. I suggest testing Envoy in its typical workload as a router.

**Comment by ztgoto** [comment]
@zhxie 
I've done tests in routing scenarios before, and the results have been published above, and the performance gap is relatively large

**Comment by philippeboyd** [comment]
Hi @howardjohn, what is the `benchtool` CLI that you used? Not sure which program it's referring to.

**Comment by jmarantz** [comment]
A few notes:
 * Envoy team has published https://www.envoyproxy.io/docs/envoy/latest/faq/performance/how_to_benchmark_envoy -- please read!
 * Envoy has a native benchmarking tool called Nighthawk, which we prefer to wrk and other tools for a variety of reasons: https://github.com/envoyproxy/nighthawk

One thing I noticed is that you specify concurrency=1 for Envoy, so it will allocate only one worker thread. In contrast your nginx config specifies 1 process. I am a little unclear on nginx's default behavior, but I think it does have a threadpool.

https://serverfault.com/questions/1098107/what-is-the-ideal-value-for-threads-on-thread-pool-in-nginx-config

Both nginx and envoy will use async i/o to multiplex many requests over a single thread, but I'm not sure how many threads nginx is using in your benchmark. Envoy will mostly use 1 (in addition to the 'main' thread for handling admin requests and config updates).

**Comment by howardjohn** [comment]
> Hi @howardjohn, what is the `benchtool` CLI that you used? Not sure which program it's referring to.

`docker run --rm --init -it --network=host howardjohn/benchtool` -- just a small wrapper around fortio basically, nothing too fancy

**Comment by ztgoto** [comment]
My understanding is that if Nginx uses multi-threading when the worker_processes is set to 1, then the CPU monitoring of the Linux system should exceed 100%. Of course, I'm not sure if my understanding is correct.

**Comment by jmarantz** [comment]
https://www.f5.com/company/blog/nginx/thread-pools-boost-performance-9x

**Comment by howardjohn** [comment]
We don't really need to compare to Nginx though -- Envoy from a few years back has the same threading/process model and is ~2x faster...

**Comment by jmarantz** [comment]
FWIW my team's product does our own benchmarking of our deployment and we have not seen that degradation.

**Comment by jmarantz** [comment]
I'm also wondering about test methdology. Can we repro the older test results from saved docker packages?

**Comment by howardjohn** [comment]
> I'm also wondering about test methdology. Can we repro the older test results from saved docker packages?

Yes -- see https://github.com/envoyproxy/envoy/issues/19103#issuecomment-2206785864. I reproduced it today again as well and see the same results.

Note config.yaml is the one from the original issue. 

Here are some more results with nighthawk as a client FWIW, similar trend

```
DEST        CLIENT     QPS    CONS  DUR  PAYLOAD  SUCCESS  THROUGHPUT   P50      P90      P99
# 30k fixed QPS
envoy-1.13  nighthawk  30000  16    10   0        299999   30000.00qps  0.029ms  0.065ms  0.475ms
envoy-1.32  nighthawk  30000  16    10   0        299997   30000.00qps  0.090ms  0.277ms  0.661ms
# 60k fixed QPS. Note we fail to hit the target with 1.32 so latency should be ignored
envoy-1.13  nighthawk  60000  16    10   0        599997   59999.93qps  0.040ms  0.222ms  0.389ms
envoy-1.32  nighthawk  60000  16    10   0        438224   43823.96qps  0.338ms  0.408ms  0.638ms
# 20k fixed QPS with larger payloads
envoy-1.13  nighthawk  20000  16    10   1024     199999   19999.99qps  0.029ms  0.049ms  0.354ms
envoy-1.32  nighthawk  20000  16    10   1024     199999   19999.98qps  0.056ms  0.143ms  0.793ms
```

**Comment by wbpcode** [comment]
> https://www.f5.com/company/blog/nginx/thread-pools-boost-performance-9x

This is a feature I never used. haha

**Comment by wbpcode** [comment]
> > I'm also wondering about test methdology. Can we repro the older test results from saved docker packages?
> 
> Yes -- see [#19103 (comment)](https://github.com/envoyproxy/envoy/issues/19103#issuecomment-2206785864). I reproduced it today again as well and see the same results.
> 
> Note config.yaml is the one from the original issue.
> 
> Here are some more results with nighthawk as a client FWIW, similar trend
> 
> ```
> DEST        CLIENT     QPS    CONS  DUR  PAYLOAD  SUCCESS  THROUGHPUT   P50      P90      P99
> # 30k fixed QPS
> envoy-1.13  nighthawk  30000  16    10   0        299999   30000.00qps  0.029ms  0.065ms  0.475ms
> envoy-1.32  nighthawk  30000  16    10   0        299997   30000.00qps  0.090ms  0.277ms  0.661ms
> # 60k fixed QPS. Note we fail to hit the target with 1.32 so latency should be ignored
> envoy-1.13  nighthawk  60000  16    10   0        599997   59999.93qps  0.040ms  0.222ms  0.389ms
> envoy-1.32  nighthawk  60000  16    10   0        438224   43823.96qps  0.338ms  0.408ms  0.638ms
> # 20k fixed QPS with larger payloads
> envoy-1.13  nighthawk  20000  16    10   1024     199999   19999.99qps  0.029ms  0.049ms  0.354ms
> envoy-1.32  nighthawk  20000  16    10   1024     199999   19999.98qps  0.056ms  0.143ms  0.793ms
> ```
> 
> (note: fully acknowledge there are more robust ways to measure performance, but at the differences we are talking about I think its fair to say there is a substantial change)

It's no doubt the change is exist and it's pretty hard to to optimize it back. Orz.

**Comment by triplewy** [comment]
Something else I've noticed about Envoy (v.1.30) is that it tends to completely collapse under high load (i.e. success rate suddenly drops to 0% and oscillates between 100% and 0% success rate every 10 minutes) whereas Nginx will have an increased response time but the success rate will be little impacted. We had to completely rollback our migration from Nginx to Envoy due to this reason. 

Another issue we noticed is that in some locations with poor network performance, Envoy's average CPU utilization would only reach 85% before they it began to collapse. This does not happen with Nginx. 

We are actually ok with Envoy potentially using 10-20% more CPU than Nginx but its tendency to completely collapse under high load made us give up our Envoy adoption.

**Comment by jmarantz** [comment]
@triplewy would you be able to take some CPU flame-graphs to see where the system is bottlenecking? Have you looked at watchdog timouts stats, or epoll histograms?

One possible fail-point in Envoy -- this is true of nginx also IIUC -- is that threads are precious and if anything being called  in the data plane unexpectedly blocks that can cause problems.

One thing that comes to mind is access logs. Do you have those enabled? What format are you using?

**Comment by krajshiva** [comment]
How is the memory situation under load? Just checking if you have Load Shed
Points enabled? (
https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/overload_manager/overload_manager#load-shed-points
).
Debug/trace level logs can throw insights into reason for a request failure.

**Comment by triplewy** [comment]
@jmarantz I re-deployed Envoy to a few of our edge clusters and was able to reproduce the oscillation behavior. What seems to be happening is:

1. Significant spike in downstream active connections causes CPU spike
2. Higher CPU load leads to higher upstream response time
3. Higher upstream response time leads to upstream pending requests
4. Upstream pending requests leads to spike in new upstream connections (We only use H1 for upstream)
5. _**Upstream connections are immediately closed after they are created**_
6. Upstream connection churn leads to failed connections and continued increasing upstream RT
7. Elevated upstream RT causes downstream clients to send resets, leading to closed upstream connections and drop in success rate.
8. Downstream active connections drops, Envoy recovers, and process repeats.

![Image](https://github.com/user-attachments/assets/d2f3806c-5967-46b8-ad11-25c2e335c94d)

During all of this, the event dispatcher and watchdog stats are relatively stable at normal levels. The concerning aspects of this are:
1. Envoy's CPU utilization is unable to remain consistently high, preventing our auto-scaler from increasing the instance count. We could use another stat to autoscale but this doesn't enforce Envoy to fully utilize its allocated resources.
2. Envoy tends to completely collapse rather than slow down processing time. The driving force behind this seems to be the immediate closing of new upstream connections but we are unsure why this happens. Using a circuit breaker to prevent excessive upstream connections simply exacerbates the problem since requests are still pending.

If we can figure out why upstream connections are immediately closed in these situations, we may be able to completely prevent Envoy from collapsing.

Below is a CPU flamegraph of our Envoy instance during non-busy time. We have access logs enabled but so does our Nginx setup.
![Image](https://github.com/user-attachments/assets/1dc457b6-6048-457b-8736-b019f7d21aeb)

**Comment by triplewy** [comment]
@jmarantz An update on our deployment: We figured out the root cause was that [max_connections_to_accept_per_socket_event](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto) is by default set to accept all pending connections from the kernel queue. For context, on Nginx it's set to accept [only one by default](https://nginx.org/en/docs/ngx_core_module.html#multi_accept). After we set `max_connections_to_accept_per_socket_event` to 1, our Envoy crashing/stalling problems were resolved. I recommend Envoy to set this to 1 by default as well. It's a very tricky configuration item to catch and it's not intuitive to configure from simply looking at event dispatcher stats. Furthermore, setting the value to 1 serves as an incredibly effective mechanism for rate-limiting during high load, I would say moreso than the overload manager itself.

**Comment by botengyao** [comment]
Thanks @triplewy for noticing this config, agreed this should be limited for overload management, and this is also limited in our environment. I will update the doc to reflect it.

**Comment by kyessenov** [comment]
What's the theory behind multi-accept causing performance degradation? Is it because a worker does more work per event and a lot of that work is wasted (due to slow upstreams)? What's the kernel queue size (128?)?

**Comment by botengyao** [comment]
I think the multi-accept could cause issues under high load, not the general performance degradation concern per versions in this issue. From an internal load test and observation, Envoy can accept >1000 connections per event from kernel, and that will use more CPU and memory at that time to let Envoy be unstable during traffic spike, e.g., overload manager cannot kick in and so one.

**Comment by triplewy** [comment]
@kyessenov Essentially Envoy gets stuck performing SSL calculations when it accepts all downstream connections at the same time. This prevents upstream requests from completing in a certain amount of time because the event loop is filled with SSL calls.

In production without setting max downstream conns to accept, we saw behavior where an influx of downstream https connections would cause our fleet of proxies to essentially collapse where all success rate dropped to 0% for 5 minutes and then it restored since the clients left from their requests were being reset. But due to our client behavior, they would come back after another 10 minutes and we'd have this 100-0% success rate oscillation during peak times in some clusters.

**Comment by kyessenov** [comment]
@triplewy OK, that makes sense. So this is really about flow control with respect to CPU, and we want reasonable fairness between upstream and downstream work.

**Comment by johnlanni** [comment]
> [@kyessenov](https://github.com/kyessenov) Essentially Envoy gets stuck performing SSL calculations when it accepts all downstream connections at the same time. This prevents upstream requests from completing in a certain amount of time because the event loop is filled with SSL calls. Our event loop duration during influxes of https connections could sometimes get to 10-30s.
> 
> In production without setting max downstream conns to accept, we saw behavior where an influx of downstream https connections would cause our fleet of proxies to essentially collapse where all success rate dropped to 0% for 5 minutes and then it restored since the clients left from their requests were being reset. But due to our client behavior, they would come back after another 10 minutes and we'd have this 100-0% success rate oscillation during peak times in some clusters.

But this should not cause Envoy to crash, right? I understand it crashed due to an OOM. The reason for the OOM, I guess, is that after accepting a large backlog of connections in one event loop, the data sent over the connections was not processed in time, leading to the TCP read buffer occupying too much memory.

**Comment by KBaichoo** [comment]
> The reason for the OOM, I guess, is that after accepting a large backlog of connections in one event loop, the data sent over the connections was not processed in time, leading to the TCP read buffer occupying too much memory.

Might be due to what you have in your listener filter chain which gets allocated on connection acceptance. I would be surprised if the TCP read buffer should be that large given that the application had not yet even accepted the connection.

**Comment by johnlanni** [comment]
> Might be due to what you have in your listener filter chain which gets allocated on connection acceptance. I would be surprised if the TCP read buffer should be that large given that the application had not yet even accepted the connection.

Yes, resource allocation related to connections on the Envoy side can also lead to OOM. Additionally, my previous statement was not clear; I did not mean that the TCP read buffer for unaccepted connections is large. What I wanted to point out is that after accepting a large number of connections in an event loop, if the data from each connection cannot be read in time, it will cause the data sent by the client on the already accepted connections to accumulate. If each event loop only accepts one request, this problem would not occur, effectively throttling the data sent by the client(limited by the maximum size of the accept queue.).

**Comment by triplewy** [comment]
@johnlanni But the tcp buffer memory is in the kernel right? That would not impact the application from OOMing. From my understanding the connections are already established from the kernel side. The application just accepts them and in this case, Envoy reads the TLS CHLO from the TCP buffer of each accepted connection and signs the SSL handshake. I think the creation of a bunch of listener filters at the same time causes the proxy to OOM.

What we think would be an interesting problem would be to see if making max_conns_to_accept_per_socket_event could be less than 1. For example 0.5. In what case should that number be lower than 1? What would you need to measure to find the optimal conns to accept for socket event (event loop).

**Comment by johnlanni** [comment]
Oh, indeed, unless in scenarios like K8s containers, where it might be OOM killed by the kubelet.

**Comment by kimiyasharifi** [comment]
@triplewy @kyessenov @botengyao We recently experienced a very similar failure pattern during a traffic spike, and I wanted to share additional data points that may help others. Even though we explicitly set:
`max_connections_to_accept_per_socket_event: 1`
we still observed envoy OOMing under heavy load. Context from our setup:

Very high spike on a single authority
Two response flags showed strong spikes:
- NoHealthyUpstream
- UpstreamOverflow

What we observed:

- During the spike, downstream request rate suddenly exceeded normal levels by a large margin (screenshots attached).
- Despite limiting accept-per-event to 1, envoy still accumulated large numbers of downstream connections quickly enough to push memory usage past limits.
- All envoy nodes eventually hit oom and were restarted by kubernetes.
- Upstream request latency increased, leading to overload behavior → pending requests → retries → more memory pressure.
- The collapse pattern matched what others described: success rate dropped to ~0% for a period, recovered slightly, then collapsed again.

<img width="1403" height="466" alt="Image" src="https://github.com/user-attachments/assets/499c3588-1495-4b2a-9139-0cac3bf3bac1" />

<img width="703" height="423" alt="Image" src="https://github.com/user-attachments/assets/932e9c7b-29bd-49e0-b034-50fc0eccb707" />

<img width="1403" height="474" alt="Image" src="https://github.com/user-attachments/assets/f65a3b53-33f1-4c61-8415-326d540739c7" />

<img width="1403" height="474" alt="Image" src="https://github.com/user-attachments/assets/bc3b7fed-a9ff-4cdc-8ddc-ad7a6b6ddfbe" />

<img width="1403" height="474" alt="Image" src="https://github.com/user-attachments/assets/f3d9f4ed-2abb-4756-b3ba-013aa73c4512" />

<img width="1403" height="474" alt="Image" src="https://github.com/user-attachments/assets/7dd2fd7a-79ca-4b99-ad6b-065c0768c265" />

<img width="1403" height="474" alt="Image" src="https://github.com/user-attachments/assets/dc6a4526-9106-479c-8efc-401aad9389c0" />

Our hypothesis so far:

1. The connection acceptance throttling helped, but
2. Per-connection memory allocations (listener filters, tls handshake buffers, upstream pending request state)
3. Combined with slow upstream response times under load
4. Still caused envoy to accumulate enough state to oom, even without multi-accept behavior.

It seems that under extreme spikes, the accept throttling alone is not sufficient to prevent memory growth from pending upstream requests + connection-level allocations.

We’re still analyzing memory profiles and would love any guidance on:

- whether additional configuration (circuit breakers? overload manager thresholds? tcp buffer tuning?) is recommended in conjunction with max_connections_to_accept_per_socket_event
- and whether there are known high-memory paths triggered during upstream overflow conditions.

If useful, I can share additional metrics or narrower traces.

**Comment by botengyao** [comment]
It seems this issue is originally on the general performance problem version by version, and the topic is diverged to overload management. And it is a really tricky system problem in general that some need to be tuned, @kimiyasharifi, some insights:

*  Do you config the overload manager and what is your config? Is the traffic mainly H/1 or H/2?
*  Consider enabling load shed points, `envoy.load_shed_points.tcp_listener_accept` for `envoy.load_shed_points.http_connection_manager_decode_headers` downstream connections and requests.
* Consider enabling load shed points `envoy.load_shed_points.connection_pool_new_connection` to prevent more connections for upstream when the system is under pressure.

  > pending requests → retries → more memory pressure
* Enabling the cluster level maximum active retries sounds a reasonable config to enable if you see retry storm.

**Comment by KBaichoo** [comment]
>It seems this issue is originally on the general performance problem version by version, and the topic is diverged to overload management. And it is a really tricky system problem in general that some need to be tuned

@kimiyasharifi do you mind raising either another issue or bringing this up in the community call: https://docs.google.com/document/d/1Yc4zkV-A_cC_R3C0u6D15WQAsRSxADs8p2l8UMMa4P4/edit?tab=t.0#heading=h.vgkvn1hr5uor to keep this thread on track with general performance?

The community call might be higher bandwidth for getting some help

**Comment by kimiyasharifi** [comment]
> It seems this issue is originally on the general performance problem version by version, and the topic is diverged to overload management. And it is a really tricky system problem in general that some need to be tuned, [@kimiyasharifi](https://github.com/kimiyasharifi), some insights:

Sorry for the delay. We haven’t enabled the overload manager in Contour, and most of our traffic is HTTP/2.

**Comment by moozechen** [comment]
Really interesting data. We've been benchmarking Envoy as a gRPC-web proxy for streaming workloads and seeing similar patterns — the per-connection overhead dominates at low concurrency but Envoy scales linearly once past ~100 concurrent streams.

One observation that might help: we found that enabling HTTP/2 priority hints (via gRPC metadata on the client side) significantly improves tail latency under mixed workloads (large file transfers + small RPCs sharing the same connection). Without priority, a single large response can add 50-200ms to concurrent small RPCs. With priority, the max tail latency impact drops to <20ms.

Also curious if you've tested with gRPC bidirectional streaming specifically? We noticed a ~15% throughput hit vs unary/server-streaming under Envoy, which seems to be from the extra buffering in the H2 frame layer. Would love to see that added to the benchmark matrix.
